### PR TITLE
Add varimp() to mojo

### PR DESCRIPTION
**Pull request by valenad1**
Mojo models supporting variable importance can call model.varimp() and varimp(int n)

Gives opportunity to replace:

```
GbmMojoModel mojo = (GbmMojoModel) ...;
VariableImportances varImps = ((SharedTreeModelAttributes) mojo._modelAttributes).getVariableImportances();
```

with:

```
MojoModel modelMojo = MojoModel.load("GBM_model.zip", true);
EasyPredictModelWrapper.Config config = new EasyPredictModelWrapper.Config().setModel(modelMojo);
EasyPredictModelWrapper model = new EasyPredictModelWrapper(config);
model.varimp();
```

and possibility to sort and get only top N variables.

Failed R tests are already fixed in rel-zemelo.

### CODE REVIEW

File: h2o-genmodel/src/main/java/hex/genmodel/attributes/parameters/IVariableImportancesHolder.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package hex.genmodel.attributes.parameters;
+
+import hex.genmodel.attributes.VariableImportances;
+
+public interface IVariableImportancesHolder {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/attributes/parameters/IVariableImportancesHolder.java:5]: I dont think we follow the convention of prefixing interfaces with `I`

File: h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java [hunk 2]
```diff
@@ -36,4 +41,27 @@ protected static VariableImportances extractFromJson(final JsonObject modelJson)
...
+     *
+     * @param n how many variables is in the output. If n >= number of variables or n <= 0 then all variables are returned.
+     * @return descending sorted array of String -> double.
+     *          Where String is variable and double is relative importance of the variable.
+     */
+    public KeyValue[] topN(int n) {
+        if (n <= 0 || n > _importances.length) {
+            n = _importances.length;
+        }
+        KeyValue[] sortedImportances = new KeyValue[_importances.length];
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java:55]: can be final

File: h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java [hunk 3]
```diff
@@ -36,4 +41,27 @@ protected static VariableImportances extractFromJson(final JsonObject modelJson)
...
+        for (int i = 0; i < _importances.length; i++) {
+            sortedImportances[i] = new KeyValue(_variables[i], _importances[i]);
+        }
+        Arrays.sort(sortedImportances, new Comparator<KeyValue>() {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java:59]: why not use lambda syntax?

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 4]
```diff
@@ -860,6 +864,34 @@ public KLimeModelPrediction predictKLime(RowData data) throws PredictException {
...
+   * See {@link VariableImportances#topN(int)}
+   */
+  public KeyValue[] varimp(int n) {
+    VariableImportances variableImportances;
+    if (m instanceof MojoModel) {
+      ModelAttributes attributes = ((MojoModel) m)._modelAttributes;
+      if (attributes == null) {
+        throw new IllegalStateException("Model attributes are not available. Did you load metadata from model? MojoModel.load(\"model\", true)");
+      } else if (attributes instanceof IVariableImportancesHolder) {
+        variableImportances = ((IVariableImportancesHolder) attributes).getVariableImportances();
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java:885]: why not just return here?

Comment (valenad1) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java:885]: True

File: h2o-genmodel/src/main/java/hex/genmodel/attributes/parameters/IVariableImportancesHolder.java (hunk 1 above)
Comment (valenad1) [reply h2o-genmodel/src/main/java/hex/genmodel/attributes/parameters/IVariableImportancesHolder.java:5]: Renamed to VariableImportancesHolder

File: h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java (hunk 3 above)
Comment (valenad1) [reply h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java:59]: Lambda syntax would be my first choice too but commend in build.gradle says it has to stay to stay Java 7 compatible:

https://github.com/h2oai/h2o-3/blob/6e4341278470bc544b9f6c950a42e146e83a3dc0/h2o-genmodel/build.gradle#L6

Comment (honzasterba) [reply h2o-genmodel/src/main/java/hex/genmodel/attributes/VariableImportances.java:59]: you are right, I forgot about this

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 5]
```diff
@@ -860,6 +864,32 @@ public KLimeModelPrediction predictKLime(RowData data) throws PredictException {
...
+      ModelAttributes attributes = ((MojoModel) m)._modelAttributes;
+      if (attributes == null) {
+        throw new IllegalStateException("Model attributes are not available. Did you load metadata from model? MojoModel.load(\"model\", true)");
+      } else if (attributes instanceof VariableImportancesHolder) {
+        return ((VariableImportancesHolder) attributes).getVariableImportances().topN(n);
+      } else {
+        throw new IllegalStateException("Model does not support variable importance");
+      }
+    } else {
+      throw new IllegalStateException("Model does not support variable importance");
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java:889]: you can move the `throw new IllegalStateException("Model does not support variable importance")` outside of the if-else blocks and leave it as the last statement of the method, thus avoiding repetition

**Review by honzasterba — APPROVED** [review]
lgtm

**Review by michalkurka — APPROVED** [review]

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java (hunk 5 above)
Comment (valenad1) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: Thanks, I will wait for jenkyns and merge it.

Comment (honzasterba) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: :+1

**Pull request merged**